bitcoin/src/util/hash.rs

```diff
@@ -39,6 +39,44 @@ where
 
 /// Calculates the merkle root of an iterator of *hashes*.
 ///
+/// This fcuntion calculate the root from a merkle path when we need to derive the root from the
+/// txs list [bitcoin_merkle_root] should be used.
+///
+/// # Returns
+/// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
+/// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
+/// - `Some(merkle_root)` if length of `hashes` is greater than one.
+pub fn bitcoin_merkle_root_from_path<T, I>(mut hashes: I) -> Option<T>
+where
+    T: Hash + Encodable,
+    <T as Hash>::Engine: io::Write,
+    I: Iterator<Item = T>,
+{
+    let first = hashes.next()?;
+    let second = match hashes.next() {
+        Some(second) => second,
+        None => return Some(first),
+    };
+
+    let mut encoder = T::engine();
+    let mut root = first;
+    root.consensus_encode(&mut encoder).expect("in-memory writers don't error");
+    second.consensus_encode(&mut encoder).expect("in-memory writers don't error");
+    root = T::from_engine(encoder);
+    for hash in hashes {
+        let mut encoder = T::engine();
+        root.consensus_encode(&mut encoder).expect("in-memory writers don't error");
+        hash.consensus_encode(&mut encoder).expect("in-memory writers don't error");
```

Comment:
If I'm not mistaken this can only compute the merkle root for leftmost branch of the tree because `root` always goes first. I'd expect here to be some condition selecting the order of hashing.

Reply:
OK a bool is enough or you prefer an enum?

Reply:
Definitely an enum here but naming such that it'll be obvious to the readers is hard.

Reply:
I suggest using
```
enum MerkleNode<H: Hash> {
    LChild(H),
    RChild(H),
}
```
and taking an array of these. I'm not sure how leftness/rightness is normally encoded (in stratum every node is implicitly a left child since we only ever prove the coinbase tx which is always first) but we should maybe impl `Encodable`/`Decodable` and/or serde `Deserialize`/`Serialize` as appropriate.

Reply:
I like that naming! I was thinking more along `enum MerkleNode { LChild, RChild }` and `(MerkleNode, H)` which simplifies the enum but I don't strongly oppose your suggestion. I'm still not convince about hash being anything else than sha256. (Note that we should be always able to make it more generic without breaking backcompat.)

Reply:
Now the function can fail cause an empty `hashes` and cause a malformed `hashes` (when left and right child are not in the right order) do we want return a Result or I stick with Option?

Reply:
eg if I have [right,left,left] I expect the function to fail
in general I always expect to have a certain numbers of left child followed by another amount of right child

Reply:
Why? Also, how are you distinguishing between "left" children and "right" children?

Reply:
```    
                                                         n0
                                              n1                     n2
                                        n3          n4         n5         n6
                                    n7     n8    n9   na    nb    nc   ne    nf
```

let's say that we want to see if nb is in n0 I expect to receive [left(n1),left(nb),right(nc),right(n2)]
result should be Some(n0).

Is easy to see that we can not have a path that start with a left child. Maybe I misunderstood the requirements

Reply:
this is the function signature:
```rust
pub fn bitcoin_merkle_root_from_path<T, I>(mut hashes: I) -> Option<T>
where
    T: Hash + Encodable,
    <T as Hash>::Engine: io::Write,
    I: Iterator<Item = MerkleChild<T>>,
{
```

Reply:
I still don't see any problem. Also maybe the proof should start from the bottom?

Reply:
io::Write seems to be necessary to use `.consensus_encode` .

About the passed path what if it is [left,right,,right,left] how I'm expected to reduce it?

Reply:
@Fi3 it sounds like in all your examples the path will simply not verify if the parity of the nodes is incorrect.

Reply:
Ok, what I'm asking is if we want to return a `Result` with a specific error eg: `VoidPath`, `ParsingError(InvalidNode)` ... or just return always a `None`.

What the function do is take an iterator of tree's nodes of type `MerkleChild` the `MerkleChild` contain an hash and can be left of right with respect to the parent node.

Then I iter through it, I expect to find at least one left child. At the first right child that I find I calculate hash(last_left_child, right_child) this new hash is a right child if there are no more nodes in the iterator, otherwise is a left child.

This give me an iterator with one node less then the one that I started with.

The above is done until only one node is left, or until we have an iterator of one left node followed by only right nodes or until an error is raised.

Reply:
When computing the proof you will just produce garbage proof, that's OK. When verifying the hashes will not match.

>  I expect to find at least one left child

Why? People may want to prove rightmost leaf.

Reply:
can you give me an example of rightmost leaf using a path with n0..ne from my message above'

Reply:
```
                                                 n0
                                      n1                     n2
                                n3          n4         n5         n6
                            n7     n8    n9   na    nb    nc   nd    ne
```

Reply:
if I want to verify ne I just pass `[left(n1), left(n5), left(nd), right(ne)]`

Reply:
We should indeed treat the leaf specially. It typically does not appear as part of a Merkle proof since it's implicit from the object that you're proving inclusion of.

Reply:
I like removing the empty case.
Just to be sure @Kixunil  you are suggesting to pass an iterator that is the path (it contain the leaf), and the leaf to recognize when we are iterating if the hash that we have is the leaf hash? And if the leaf is not in the iterator we just return the leaf?

Reply:
No, the leaf is not in the iterator, there's no point if the argument already contains it. So the path only contains intermediate nodes.

Reply:
ok so is an iterator of:
```rust
enum MerkleNode<H: Hash> {
    LChild(H),
    RChild(H),
}
```

Reply:
if the leaf is into the iterator  we could avoid to introduce the  type `MerkleNode` as it would be obvious which are left child and which are right child, this will make the API easier to use

Reply:
If the leaf is the first element the leaf is left and every other node are right.
If the leaf is the last node the leaf is right and every other node is left.
If the leaf is in another position  every node before the leaf is left and every node after is right.

Reply:
How do you know what position the leaf is in?

Reply:
btw still not sure if we want to specify the position of all the elements into the iterator.
I could pass an iterator that is the path without and is composed by simple hashes and contains the leaf so it is likely something that as a user I do not need to change (changing the Inner types or removing the leaf)
Then I pass the leaf that need to specify the position.

Does it make sense from an usability point of view?

Reply:
Sorry, I have been conflating two things:

From a usability point of view we would like to be able to create and validate the `CMerkleBlock` structure that is output from Core's `gettxoutproof` RPC. This is actually an interesting structure which is able to prove more than one TXID at a time. I think this is what @Kixunil and I really would like.

Separately, we want to be able to validate a stratum v2 merkle proof, which is just a single list of hashes (all of which are right-parents because the proof is always about a coinbase tx). I think this is what you're trying specifically to achieve.

What may make sense is to have two completely separate APIs -- one which is called `check_stratum_v2_coinbase_merkle_proof` or something similarly crazy, and one based around the txoutproof format. Alternately, we could have a `TxOutProof` structure and add a constructor that takes a stratum v2 proof and synthetically constructs a gettxoutproof object from that.

Reply:
Why do we need a separate API? Just use `map(MerkleNode::Right)` on the existing iterator for stratum.